[Agent Builder] Document trace message content attributes and where to view traces (#7602)
elastic/docs-content: explore-analyze/ai-features/agent-builder/agent-traces-dashboard.md

| `attributes.gen_ai.request.model` | Model name |
| `attributes.gen_ai.provider.name` | Model provider |
| `attributes.gen_ai.agent.id` | Agent identifier |
| `attributes.gen_ai.conversation.id` | Conversation identifier |

> | `attributes.elastic.inference.span.kind` | The kind of work a span represents: `LLM` on `chat` spans, `TOOL` on `execute_tool` spans, and `CHAIN` or `AGENT` on `invoke_agent` spans, where `CHAIN` is a conversation round and `AGENT` is an agent execution. Internal spans such as `generate_title` also use `CHAIN`, so combine this field with a `span.name` filter instead of using it on its own |

Reviewer: I wonder if we could use line breaks or something to break this up a little, it's quite a dense chunk of text?

The dashboard does not use these fields, but you can query them yourself. When an administrator opts in to capturing conversation content, prompts, responses, and tool call content are stored in the following attributes. Each one depends on a [trace privacy setting](collect-traces.md#trace-privacy-settings), and all of those settings are off by default.

| Field | Span | Description | Requires |
|---|---|---|---|
| `attributes.gen_ai.input.messages` | `chat` | Chat history sent to the model, as a series of user, assistant, and tool turns | **Include user prompts in traces** for the user turns, **Include LLM responses in traces** for the assistant turns, and **Include tool call details in traces** for the tool turns |
| `attributes.gen_ai.output.messages` | `chat` | Model responses | **Include LLM responses in traces** |
| `attributes.gen_ai.system_instructions` | `chat` | System prompt | **Include system prompt in traces** |
| `attributes.gen_ai.tool.call.arguments` | `execute_tool` | Arguments passed to the tool | **Include tool call details in traces** |
| `attributes.gen_ai.tool.call.result` | `execute_tool` | Value the tool returned | **Include tool call details in traces** |

> Any turn that a privacy setting excludes is dropped from `attributes.gen_ai.input.messages`. When all three of the settings that govern it are off, the field is an empty array (`[]`). Filter those rows out, as the following example does. The `attributes.gen_ai.tool.call.id` field on the `execute_tool` spans is not affected by the privacy settings, though it is absent when a tool call has no id.

Reviewer: this whole section after the table is kind of a wall of text

consider subheadings (or any other technique) to break this up

**Include tool call details in traces** reaches further than the tool turns. When it is off, tool calls are also removed from the assistant turns that remain, in both `attributes.gen_ai.input.messages` and `attributes.gen_ai.output.messages`, so an assistant turn keeps its text but not the call it made.

> Spans carry other content-bearing attributes besides these. The `chat` spans record the definitions of the tools offered to the model in `attributes.gen_ai.tool.definitions`, including each tool's description and parameter schema, and the `execute_tool` spans record the tool's own description in `attributes.gen_ai.tool.description`.

Reviewer: ```suggestion
Spans also contain other content-bearing attributes. The `chat` spans record the definitions of the tools offered to the model in `attributes.gen_ai.tool.definitions`, including each tool's description and parameter schema, and the `execute_tool` spans record the tool's own description in `attributes.gen_ai.tool.description`.
```

nit: I'd replace all instances of "carry" which feels a little imprecise

Reviewer: sidebar: it would be nice to link to a canonical definition of a span for the first instance of "span" in these docs (if we have one) and perhaps have a short explanation of the anatomy of a span 

maybe it's this? https://opentelemetry.io/docs/concepts/signals/traces/#spans

> Each content field holds a JSON string rather than indexed text, following the [OpenTelemetry semantic conventions for generative AI](https://github.com/open-telemetry/semantic-conventions-genai). The message attributes hold an array of `{"role": ..., "parts": [...]}` objects, where each part is a `text`, `tool_call`, or `tool_call_response` item, and `attributes.gen_ai.system_instructions` holds an array of `{"type": ..., "content": ...}` objects.

Reviewer: there are a few spots like this where breaking some out stuff into bullets would improve the reading experience

> :::{important}

Reviewer: This is far too much information for an admonition and therefore deserves its own subsection :)

To read the full content, request the document with [{{es}} search](/solutions/search/querying-for-search.md) instead of {{esql}}, and ask for `_source`. The `fields` option applies the same limit and returns nothing.

To find the affected spans in the first place, use the `_ignored` metadata field, which lists the fields on a document that were not indexed. For these attributes, that means the value passed the length limit. It is available both on search hits and in {{esql}}, as shown in [Example queries](#example-queries).

Tool arguments and results are also recorded on the `execute_tool` spans, in `attributes.gen_ai.tool.call.arguments` and `attributes.gen_ai.tool.call.result`. Each of those holds one tool call rather than the whole conversation, so it is less likely to pass the limit, though a large tool result still can. Prefer them when you query tool activity with {{esql}}.
:::

> ### Example queries

Reviewer: nit: it would be nice, for scannability, to have a H4 heading starting with a verb that encapsulates each query

| `name` | Span name. On `execute_tool` spans it is `execute_tool <tool-id>`, for example `execute_tool platform.core.list_indices`. For the bare tool id, use `attributes.gen_ai.tool.name` |
| `duration` | Span duration in nanoseconds (root field). Divide by 1,000,000,000 for seconds |
| `status.code` | Span status, for example `Error` (root field) |
| `@timestamp` | When the span started |

### Message content attributes [message-content-attributes]

The dashboard does not use these fields, but you can query them yourself. When an administrator opts in to capturing conversation content, prompts, responses, and tool call content are stored in the following attributes. Each one depends on a [trace privacy setting](collect-traces.md#trace-privacy-settings), and all of those settings are off by default.

> | Field | Span | Description | Requires |

Reviewer: ```suggestion
| Field | Span | Description | Required setting(s) |
```

nit for clarity

- The `chat` spans record the definitions of the tools offered to the model in `attributes.gen_ai.tool.definitions` (including each tool's description and parameter schema).
- The `execute_tool` spans record the tool's own description in `attributes.gen_ai.tool.description`.

#### JSON payload structure

Each content field holds a JSON string rather than indexed text, following the [OpenTelemetry semantic conventions for generative AI](https://github.com/open-telemetry/semantic-conventions-genai):
- The message attributes hold an array of `{"role": ..., "parts": [...]}` objects, where each part is a `text`, `tool_call`, or `tool_call_response` item.
- `attributes.gen_ai.system_instructions` holds an array of `{"type": ..., "content": ...}` objects.

#### Handling large content fields

These content fields are `keyword` fields with `ignore_above` set to `1024`. A value longer than 1024 characters is not indexed, and {{esql}} returns it as `null` even though the full value is stored in the document. System prompts, model responses, and any chat history that contains a tool result routinely pass 1024 characters, so treat {{esql}} as dependable only for short values here.

> **Reading full content**

Reviewer: nit: Maybe add a newline between these 3 bold sentences and their accompanying text?
